File: examples/canonical-auth/java/src/main/java/dev/tenantcore/auth/domain/InMemoryStore.java
package dev.tenantcore.auth.domain;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import org.springframework.stereotype.Repository;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

@Repository
public class InMemoryStore {
    private static final List<String> CANONICAL_PERMISSION_NAMES = List.of(
        "iam:user:read",
        "iam:user:create",
        "iam:user:update",
        "iam:user:delete",
        "iam:user:invite",
        "iam:group:read",
        "iam:group:create",
        "iam:group:update",
        "iam:group:delete",
        "iam:group:assign_permission",
        "iam:group:assign_user",
        "iam:tenant:read",
        "iam:tenant:create",
        "iam:tenant:update",
        "iam:permission:read",
        "billing:invoice:read",
        "billing:invoice:create",
        "billing:subscription:read",
        "billing:subscription:update",
        "reports:usage:read",
        "reports:usage:export",
        "reports:audit:read",
        "admin:tenant:create",
        "admin:tenant:suspend",
        "admin:audit:read"
    );

    private final Map<String, User> users = new HashMap<>();
    private final Map<String, Tenant> tenants = new HashMap<>();
    private final Map<String, Group> groups = new HashMap<>();
    private final Map<String, Permission> permissions = new HashMap<>();

    private final List<Membership> memberships = new ArrayList<>();
    private final List<GroupPermission> groupPermissions = new ArrayList<>();
    private final List<UserGroup> userGroups = new ArrayList<>();

    public record Membership(
        String id,
        @JsonProperty("user_id") String userId,
        @JsonProperty("tenant_id") String tenantId,
        @JsonProperty("tenant_role") String tenantRole,
        @JsonProperty("created_at") Instant createdAt,
        @JsonProperty("is_active") boolean isActive
    ) {}

    public record GroupPermission(
        String id,
        @JsonProperty("group_id") String groupId,
        @JsonProperty("permission_id") String permissionId,
        @JsonProperty("granted_at") Instant grantedAt
    ) {}

    public record UserGroup(
        String id,
        @JsonProperty("user_id") String userId,
        @JsonProperty("group_id") String groupId,
        @JsonProperty("joined_at") Instant joinedAt
    ) {}

    public static InMemoryStore loadFromFixtures(Path fixtureDir) throws IOException {
        ObjectMapper mapper = new ObjectMapper()
            .registerModule(new JavaTimeModule())
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

        InMemoryStore store = new InMemoryStore();

        List<User> userList = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("users.json")), new TypeReference<>() {});
        userList.forEach(u -> store.users.put(u.id(), u));

        List<Tenant> tenantList = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("tenants.json")), new TypeReference<>() {});
        tenantList.forEach(t -> store.tenants.put(t.id(), t));

        List<Group> groupList = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("groups.json")), new TypeReference<>() {});
        groupList.forEach(g -> store.groups.put(g.id(), g));

        List<Permission> permList = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("permissions.json")), new TypeReference<>() {});
        permList.forEach(p -> store.permissions.put(p.id(), p));
        store.seedCanonicalPermissions();

        List<Membership> memList = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("memberships.json")), new TypeReference<>() {});
        store.memberships.addAll(memList);

        List<GroupPermission> gpList = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("group_permissions.json")), new TypeReference<>() {});
        store.groupPermissions.addAll(gpList);

        List<UserGroup> ugList = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("user_groups.json")), new TypeReference<>() {});
        store.userGroups.addAll(ugList);

        return store;
    }

    public User getUserById(String id) {
        return users.get(id);
    }

    public User getUserByEmail(String email) {
        return users.values().stream().filter(u -> u.email().equals(email)).findFirst().orElse(null);
    }

    public Optional<Membership> getActiveMembership(String userId) {
        return memberships.stream()
            .filter(m -> m.userId().equals(userId) && m.isActive())
            .findFirst();
    }

    public Tenant getTenantById(String id) {
        return tenants.get(id);
    }

    public String getTenantName(String id) {
        Tenant t = tenants.get(id);
        return t != null ? t.name() : null;
    }

    public List<Group> getGroupsForUser(String userId, String tenantId) {
        return userGroups.stream()
            .filter(ug -> ug.userId().equals(userId))
            .map(ug -> groups.get(ug.groupId()))
            .filter(Objects::nonNull)
            .filter(g -> tenantId == null || tenantId.equals(g.tenantId()))
            .collect(Collectors.toList());
    }

    public List<String> getEffectivePermissions(String userId) {
        Optional<Membership> membership = getActiveMembership(userId);
        if (membership.isPresent()) {
            String tenantRole = membership.get().tenantRole();
            if ("tenant_admin".equals(tenantRole)) {
                return permissions.values().stream()
                    .map(Permission::name)
                    .sorted()
                    .toList();
            }
            if ("super_admin".equals(tenantRole)) {
                return getAdminPermissions();
            }
        }

        Set<String> perms = new HashSet<>();
        for (UserGroup ug : userGroups) {
            if (ug.userId().equals(userId)) {
                for (GroupPermission gp : groupPermissions) {
                    if (gp.groupId().equals(ug.groupId())) {
                        Permission p = permissions.get(gp.permissionId());
                        if (p != null) {
                            perms.add(p.name());
                        }
                    }
                }
            }
        }
        return perms.stream().sorted().toList();
    }

    public boolean verifyMembership(String userId, String tenantId) {
        return memberships.stream()
            .anyMatch(m -> m.userId().equals(userId) && Objects.equals(m.tenantId(), tenantId) && m.isActive());
    }

    public String getMembershipRole(String userId, String tenantId) {
        return memberships.stream()
            .filter(m -> m.userId().equals(userId) && Objects.equals(m.tenantId(), tenantId) && m.isActive())
            .map(Membership::tenantRole)
            .findFirst()
            .orElse(null);
    }

    public List<String> getAdminPermissions() {
        return permissions.values().stream()
            .map(Permission::name)
            .filter(name -> name.startsWith("admin:"))
            .sorted()
            .toList();
    }

    public Collection<Permission> getAllPermissions() {
        return permissions.values().stream()
            .sorted(Comparator.comparing(Permission::name))
            .toList();
    }

    public Collection<User> getAllUsers(String tenantId) {
        return users.values().stream()
            .filter(u -> tenantId == null || Objects.equals(u.tenantId(), tenantId))
            .collect(Collectors.toList());
    }

    public Collection<Tenant> getAllTenants() {
        return tenants.values().stream()
            .sorted(Comparator.comparing(Tenant::slug))
            .toList();
    }

    public Collection<Group> getAllGroups(String tenantId) {
        return groups.values().stream()
            .filter(g -> tenantId == null || Objects.equals(g.tenantId(), tenantId))
            .collect(Collectors.toList());
    }

    public User inviteUser(String tenantId, String email, String displayName, List<String> groupSlugs) {
        User user = new User(
            "u-" + UUID.randomUUID(),
            email,
            displayName,
            tenantId,
            Instant.now(),
            true,
            1
        );
        users.put(user.id(), user);
        memberships.add(new Membership(
            "m-" + UUID.randomUUID(),
            user.id(),
            tenantId,
            "tenant_member",
            Instant.now(),
            true
        ));
        if (groupSlugs != null) {
            groupSlugs.stream()
                .map(this::findGroupBySlug)
                .filter(Objects::nonNull)
                .filter(group -> tenantId.equals(group.tenantId()))
                .forEach(group -> userGroups.add(new UserGroup(
                    "ug-" + UUID.randomUUID(),
                    user.id(),
                    group.id(),
                    Instant.now()
                )));
        }
        return user;
    }

    public Group createGroup(String tenantId, String slug, String name) {
        Group group = new Group(
            "g-" + UUID.randomUUID(),
            tenantId,
            slug,
            name,
            Instant.now()
        );
        groups.put(group.id(), group);
        return group;
    }

    public boolean assignPermissionToGroup(String groupId, String permissionName, String tenantId) {
        Group group = groups.get(groupId);
        if (group == null || !Objects.equals(group.tenantId(), tenantId)) {
            return false;
        }
        Permission permission = permissions.values().stream()
            .filter(value -> value.name().equals(permissionName))
            .findFirst()
            .orElse(null);
        if (permission == null) {
            return false;
        }
        groupPermissions.add(new GroupPermission(
            "gp-" + UUID.randomUUID(),
            groupId,
            permission.id(),
            Instant.now()
        ));
        return true;
    }

    public boolean assignUserToGroup(String groupId, String userId, String tenantId) {
        Group group = groups.get(groupId);
        User user = users.get(userId);
        if (group == null || user == null) {
            return false;
        }
        if (!Objects.equals(group.tenantId(), tenantId) || !Objects.equals(user.tenantId(), tenantId)) {
            return false;
        }
        userGroups.add(new UserGroup(
            "ug-" + UUID.randomUUID(),
            userId,
            groupId,
            Instant.now()
        ));
        return true;
    }

    public Tenant createTenant(String slug, String name, String firstAdminEmail) {
        String tenantId = "t-" + UUID.randomUUID();
        Tenant tenant = new Tenant(tenantId, slug, name, Instant.now(), true);
        tenants.put(tenant.id(), tenant);
        User admin = new User(
            "u-" + UUID.randomUUID(),
            firstAdminEmail,
            name + " Admin",
            tenant.id(),
            Instant.now(),
            true,
            1
        );
        users.put(admin.id(), admin);
        memberships.add(new Membership(
            "m-" + UUID.randomUUID(),
            admin.id(),
            tenant.id(),
            "tenant_admin",
            Instant.now(),
            true
        ));
        return tenant;
    }

    private Group findGroupBySlug(String slug) {
        return groups.values().stream()
            .filter(group -> group.slug().equals(slug))
            .findFirst()
            .orElse(null);
    }

    private void seedCanonicalPermissions() {
        Set<String> existingNames = permissions.values().stream()
            .map(Permission::name)
            .collect(Collectors.toSet());
        for (String permissionName : CANONICAL_PERMISSION_NAMES) {
            if (existingNames.contains(permissionName)) {
                continue;
            }
            String permissionId = "perm-" + permissionName.replace(':', '-');
            permissions.put(
                permissionId,
                new Permission(
                    permissionId,
                    permissionName,
                    "Canonical seeded permission",
                    Instant.parse("2025-01-01T00:00:00Z")
                )
            );
        }
    }
}

File: examples/canonical-auth/kotlin/src/main/kotlin/dev/tenantcore/auth/domain/InMemoryStore.kt
package dev.tenantcore.auth.domain

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import java.nio.file.Files
import java.nio.file.Path
import java.time.Instant
import java.util.UUID

class InMemoryStore private constructor(
    private val users: MutableMap<String, User>,
    private val tenants: MutableMap<String, Tenant>,
    private val groups: MutableMap<String, Group>,
    private val permissions: MutableMap<String, Permission>,
    private val memberships: MutableList<Membership>,
    private val groupPermissions: MutableList<GroupPermission>,
    private val userGroups: MutableList<UserGroup>,
) {
    fun getUserById(id: String): User? = users[id]

    fun getUserByEmail(email: String): User? = users.values.firstOrNull { it.email == email }

    fun getActiveMembership(userId: String): Membership? =
        memberships.firstOrNull { it.userId == userId && it.isActive }

    fun verifyMembership(userId: String, tenantId: String): Boolean =
        memberships.any { it.userId == userId && it.tenantId == tenantId && it.isActive }

    fun getTenantName(tenantId: String): String? = tenants[tenantId]?.name

    fun getGroupsForUser(userId: String, tenantId: String): List<Group> =
        userGroups.filter { it.userId == userId }
            .mapNotNull { groups[it.groupId] }
            .filter { it.tenantId == tenantId }

    fun getEffectivePermissions(userId: String): List<String> {
        val membership = getActiveMembership(userId)
        return when (membership?.tenantRole) {
            "tenant_admin" -> permissions.values.map { it.name }.sorted()
            "super_admin" -> getAdminPermissions()
            else -> {
                val groupIds = userGroups.filter { it.userId == userId }.map { it.groupId }.toSet()
                groupPermissions.filter { it.groupId in groupIds }
                    .mapNotNull { permissions[it.permissionId]?.name }
                    .toSet()
                    .sorted()
            }
        }
    }

    fun getAdminPermissions(): List<String> =
        permissions.values.map { it.name }.filter { it.startsWith("admin:") }.sorted()

    fun getAllUsers(tenantId: String): List<User> =
        users.values.filter { it.tenantId == tenantId }.sortedBy { it.email }

    fun getAllGroups(tenantId: String): List<Group> =
        groups.values.filter { it.tenantId == tenantId }.sortedBy { it.slug }

    fun getAllPermissions(): List<Permission> = permissions.values.sortedBy { it.name }

    fun getAllTenants(): List<Tenant> = tenants.values.sortedBy { it.slug }

    fun inviteUser(tenantId: String, email: String, displayName: String, groupSlugs: List<String>): User {
        val user = User(
            id = "u-${UUID.randomUUID()}",
            email = email,
            displayName = displayName,
            tenantId = tenantId,
            createdAt = Instant.now(),
            isActive = true,
            aclVer = 1,
        )
        users[user.id] = user
        memberships += Membership(
            id = "m-${UUID.randomUUID()}",
            userId = user.id,
            tenantId = tenantId,
            tenantRole = "tenant_member",
            createdAt = Instant.now(),
            isActive = true,
        )
        groupSlugs.mapNotNull { slug -> groups.values.firstOrNull { it.slug == slug && it.tenantId == tenantId } }
            .forEach { group ->
                userGroups += UserGroup(
                    id = "ug-${UUID.randomUUID()}",
                    userId = user.id,
                    groupId = group.id,
                    joinedAt = Instant.now(),
                )
            }
        return user
    }

    fun createGroup(tenantId: String, slug: String, name: String): Group {
        val group = Group(
            id = "g-${UUID.randomUUID()}",
            tenantId = tenantId,
            slug = slug,
            name = name,
            createdAt = Instant.now(),
        )
        groups[group.id] = group
        return group
    }

    fun assignPermissionToGroup(groupId: String, permissionName: String, tenantId: String): Boolean {
        val group = groups[groupId] ?: return false
        if (group.tenantId != tenantId) return false
        val permission = permissions.values.firstOrNull { it.name == permissionName } ?: return false
        groupPermissions += GroupPermission(
            id = "gp-${UUID.randomUUID()}",
            groupId = groupId,
            permissionId = permission.id,
            grantedAt = Instant.now(),
        )
        return true
    }

    fun assignUserToGroup(groupId: String, userId: String, tenantId: String): Boolean {
        val group = groups[groupId] ?: return false
        val user = users[userId] ?: return false
        if (group.tenantId != tenantId || user.tenantId != tenantId) return false
        userGroups += UserGroup(
            id = "ug-${UUID.randomUUID()}",
            userId = userId,
            groupId = groupId,
            joinedAt = Instant.now(),
        )
        return true
    }

    fun createTenant(slug: String, name: String, firstAdminEmail: String): Tenant {
        val tenant = Tenant(
            id = "t-${UUID.randomUUID()}",
            slug = slug,
            name = name,
            createdAt = Instant.now(),
            isActive = true,
        )
        tenants[tenant.id] = tenant
        val admin = User(
            id = "u-${UUID.randomUUID()}",
            email = firstAdminEmail,
            displayName = "$name Admin",
            tenantId = tenant.id,
            createdAt = Instant.now(),
            isActive = true,
            aclVer = 1,
        )
        users[admin.id] = admin
        memberships += Membership(
            id = "m-${UUID.randomUUID()}",
            userId = admin.id,
            tenantId = tenant.id,
            tenantRole = "tenant_admin",
            createdAt = Instant.now(),
            isActive = true,
        )
        return tenant
    }

    companion object {
        private val canonicalPermissionNames = listOf(
            "iam:user:read",
            "iam:user:create",
            "iam:user:update",
            "iam:user:delete",
            "iam:user:invite",
            "iam:group:read",
            "iam:group:create",
            "iam:group:update",
            "iam:group:delete",
            "iam:group:assign_permission",
            "iam:group:assign_user",
            "iam:tenant:read",
            "iam:tenant:create",
            "iam:tenant:update",
            "iam:permission:read",
            "billing:invoice:read",
            "billing:invoice:create",
            "billing:subscription:read",
            "billing:subscription:update",
            "reports:usage:read",
            "reports:usage:export",
            "reports:audit:read",
            "admin:tenant:create",
            "admin:tenant:suspend",
            "admin:audit:read",
        )

        fun loadFromFixtures(fixtureDir: Path): InMemoryStore {
            val mapper = ObjectMapper()
                .registerKotlinModule()
                .registerModule(JavaTimeModule())
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            val users = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("users.json")), object : TypeReference<List<User>>() {})
            val tenants = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("tenants.json")), object : TypeReference<List<Tenant>>() {})
            val groups = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("groups.json")), object : TypeReference<List<Group>>() {})
            val permissions = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("permissions.json")), object : TypeReference<List<Permission>>() {})
            val memberships = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("memberships.json")), object : TypeReference<List<Membership>>() {})
            val groupPermissions = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("group_permissions.json")), object : TypeReference<List<GroupPermission>>() {})
            val userGroups = mapper.readValue(Files.readAllBytes(fixtureDir.resolve("user_groups.json")), object : TypeReference<List<UserGroup>>() {})

            val store = InMemoryStore(
                users.associateBy { it.id }.toMutableMap(),
                tenants.associateBy { it.id }.toMutableMap(),
                groups.associateBy { it.id }.toMutableMap(),
                permissions.associateBy { it.id }.toMutableMap(),
                memberships.toMutableList(),
                groupPermissions.toMutableList(),
                userGroups.toMutableList(),
            )
            store.seedCanonicalPermissions()
            return store
        }
    }

    private fun seedCanonicalPermissions() {
        val existing = permissions.values.map { it.name }.toSet()
        canonicalPermissionNames.filterNot(existing::contains).forEach { permissionName ->
            val permissionId = "perm-${permissionName.replace(':', '-')}"
            permissions[permissionId] = Permission(
                id = permissionId,
                name = permissionName,
                description = "Canonical seeded permission",
                createdAt = Instant.parse("2025-01-01T00:00:00Z"),
            )
        }
    }
}
